# **STAGE 1 - ETL**

## **Objectives**

In this Jupyter Notebook, I will be following the ETL process on the raw dataset: `student_performance_dataset.csv`

The process will be as follows:
* **Extract**: I will load the dataset into a pandas DataFrame and perform an initial inspection of its structure, data types and quality.
* **Transform**: I will clean and standardise the dataset, change data types where necessary, remove any null values, and engineer new features.
* **Load**: Finally, I will export the cleaned DataFrame as a new CSV file, ready to be used in the next Jupyter Notebook - **Stage 2 - Visualisation**.

## **Inputs**

* Write down which data or information you need to run the notebook 

## **Outputs**

* Write here which files, code or artefacts you generate by the end of the notebook 

## **Additional Comments**

* If you have any additional comments that don't fit in the previous bullets, please state them here. 

---

## **Change working directory**

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'/Users/elliebrawn/Documents/vscode-projects/student-performance-analysis/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'/Users/elliebrawn/Documents/vscode-projects/student-performance-analysis'

---

## **Import Libraries and Packages**

In order to read `student_performance_dataset.csv` and create visualisations for exploratory analysis, I need to install the following libraries and packages:

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style("whitegrid")

---

## **Core Statistical Concepts**

Before diving into this project, I am first going to outline some key statistical concepts that are foundational to data analysis. Because many of these will appear in my project, it is important to outline what they mean and how they are used.

I'll split these up into three main sections: 

### **Descriptive Statistics**
Descriptive statistics summarise features of a data set. They can describe things like the **central tendency** and the **variability**.

| **Concept** | **Sub-Concept** | **Description** |
| ---------- | --------------| -------- |
| **Central Tendency** | - | The central tendency refers to a value that shows the middle or typical point in a dataset. |
| - | **Mean** | The mean is the average value of a dataset. <br><br>***How to Calculate***: Add all the values together and divide by the total number of observations. <br><br>***Pandas method to calculate***: Use `.mean()`. <br><br> The mean provides a simple measure of the central tendency of the data, although it can be strongly affected by unusually large or small values (outliers). |
| - | **Median** | The median is the middle value - it can be thought of as the element that divides the data into two equal parts. <br><br>***How to Calculate***: Arrange the data in ascending or descending order. If the number of elements is odd, the median is the middle element in the arranged values. If the number of observations is even, the median is given by the mean of the two middle elements. <br><br>***Pandas method to calculate***: Use `.median()`. <br><br>Unlike the mean, the median is less affected by outliers, making it useful for datasets with skewed distributions. |
| - | **Mode** | The mode is the value that appears most often (has the highest frequency) in a dataset. <br><br>***How to Calculate***: The easiest way is to order the numbers in ascending or descending order. Then, count the number of times a value appears. <br><br>***Pandas method to Calculate***: Use `.mode()`. <br><br>A dataset can be **Uni-modal** (only one mode), **Bi-modal** (two modes), or **Multi-modal** (more than two modes). |
| **Variability** | - | Measures of variability are also known as dispersion or spread and show how spread out the data is. Some of the ways to measure this are **standard deviation**, **range** and **variance**. |
| - | **Standard Deviation** | Standard deviation measures how spread out the observations are around the mean. It is used by data analysts to assess the distribution of the data. A small standard deviation means values tend to be close to the mean. A large standard deviation indicates greater variability. |

### **Inferential Statistics**
Whereas descriptive statistics summarise the data you have, inferential statistics use data to draw conclusions beyond the observed sample.

| **Concept** | **Sub-Concept** | **Description** |
| ---------- | --------------| -------- |
| **Hypothesis Testing** | - | This is a statistical method used to determine whether there is sufficient evidence to support a claim. <br><br> A **null hypothesis (H0)** represents no effect or difference, and an **alternative hypothesis (Ha or H1)**, which represents an effect or difference. A statistical test and a significance level are then used to determine whether the evidence is strong enough to reject the null hypothesis. |

### **Probability**
Statistical probability measures how likely an event is to occur and is expressed as a number between 0 and 1.

| **Concept** | **Sub-Concept** | **Description** |
| ---------- | --------------| -------- |
| **p-value** | - | The p-value is a statistical measure and specific probability that tells you the likelihood that an observed outcome is the result of chance or that it could have occurred under the null hypothesis. It tells you whether a result is **statistically significant**. |

### **Why are these core concepts so important in Data Analysis?**
All of the concepts listed together are the foundation of data analysis. They help to transform raw data into meaningful information and help to support evidence-based decision-making.

---

## **1. Extraction**

In this first section, Extraction, I am looking to gain a baseline understanding of what my data looks like before I move on to cleaning and transforming.

I am using the acronym - **DISH** - to guide this process, where the letters stand for:

| Letter | Function / Attribute |
| ------ | ------- |
|**D**| `.describe`|
|**I**| `.info()`|
|**S**| `.shape`|
|**H**| `.head()`|

I will also use the following extraction steps:

| Function / Attribute |
| -----|
|`.dtypes`|
|`.isnull.sum()`|

### **1.1 Load the raw dataset**

Firstly, I will load the raw dataset - `student_performance_dataset.csv` - that I am going to be working on for this ETL Jupyter Notebook.

In [5]:
# Load the raw dataset
df_raw = pd.read_csv("datasets/raw-data/student_performance_dataset.csv")

# Display the first few rows of the dataset to make sure the data has loaded correctly
df_raw.head()

,student_id,gender,study_time_hours,attendance_percent,sleep_hours,parental_education,internet_access,extracurricular_activities,part_time_job,previous_grade,final_exam_score,final_grade
0,1,Male,4.0,98.0,6.5,Bachelors,Yes,Yes,No,76.9,100.0,A
1,2,Female,6.3,100.0,5.7,High School,Yes,Yes,Yes,75.5,100.0,A
2,3,Male,4.9,85.3,7.9,Bachelors,Yes,No,Yes,88.5,97.3,A
3,4,Male,2.6,77.5,8.0,NaN,Yes,Yes,No,85.1,83.8,B
4,5,Male,2.2,89.6,4.6,Bachelors,Yes,No,Yes,61.8,68.3,D


### **1.2 Display the shape of the dataset**

In [6]:
# Display the shape of the dataset
print(f"Shape of the dataset: {df_raw.shape}")

Shape of the dataset: (1000, 12)


The dataset has 1,000 rows and 12 columns.

### **1.3 Display the data types for each column**

Now, I am going to inspect the data types for each column of my dataframe.

In [7]:
# Outline the data types of each column in the dataset
print(f"Data types of each column:\n{df_raw.dtypes}")

Data types of each column:
student_id                      int64
gender                         object
study_time_hours              float64
attendance_percent            float64
sleep_hours                   float64
parental_education             object
internet_access                object
extracurricular_activities     object
part_time_job                  object
previous_grade                float64
final_exam_score              float64
final_grade                    object
dtype: object


### **1.4 Display a summary of all columns in the dataset**

In [8]:
# Use .info() to get a concise summary of the dataset, including the number of non-null entries
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   student_id                  1000 non-null   int64  
 1   gender                      1000 non-null   object 
 2   study_time_hours            1000 non-null   float64
 3   attendance_percent          1000 non-null   float64
 4   sleep_hours                 1000 non-null   float64
 5   parental_education          898 non-null    object 
 6   internet_access             1000 non-null   object 
 7   extracurricular_activities  1000 non-null   object 
 8   part_time_job               1000 non-null   object 
 9   previous_grade              1000 non-null   float64
 10  final_exam_score            1000 non-null   float64
 11  final_grade                 1000 non-null   object 
dtypes: float64(5), int64(1), object(6)
memory usage: 93.9+ KB


This method builds on what we have already established using `.shape` and `.dtypes`. 

It shows us that we have 1000 entries and 12 data columns and the data types for each of these columns. 

Additionally, we can see an insight into `Non-Null Count`, aka the number of values in each column that aren't null. 
* I will explore this further down in **1.6 Check for Missing Values in the Dataset**. 
* This summary of the dataset clearly shows that `parental_education` has missing values - `898 non-null` vs the expected `1000 non-null`.

### **1.5 Display summary statistics for the dataset**

In [9]:
# Use .describe() for summary statistics of the numerical columns in the dataset
df_raw.describe()

,student_id,study_time_hours,attendance_percent,sleep_hours,previous_grade,final_exam_score
count,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000
mean,500.500000,3.570700,85.092300,6.799500,69.740900,83.543500
std,288.819436,1.478559,9.270685,1.203527,12.613425,10.341333
min,1.000000,0.500000,54.800000,3.200000,31.300000,46.800000
25%,250.750000,2.600000,78.800000,5.900000,61.000000,76.075000
50%,500.500000,3.600000,85.200000,6.800000,69.600000,83.800000
75%,750.250000,4.500000,91.900000,7.600000,78.400000,91.525000
max,1000.000000,8.100000,100.000000,10.000000,100.000000,100.000000


As you can see, `.describe()` shows some of the core descriptive statistics measures that I have described in more detail in **Core Statistical Concepts**. 

The `student_id` column shows a high standard deviation, meaning variability in the data is high. If we look again at first five rows of the dataset (using `.head()` above), we understand why: the values in `student_id` are `1`, `2`, `3`, `4`, `5...` and looking at the  max figure above, it goes all the way up to `1000`. These integers are acting like labels for each student and therefore this variability is explainable. 

There are no minus numbers included in any of the numerical columns and with a quick look over all of the figures, nothing seems particularly anomalous. I will look at the interquartile range and outlier detection in **2. Transform** if any outliers need handling.

#### **1.6 Check for Missing Values in the Dataset**

In [10]:
## Check for null values in the dataset
print(f"Number of null values in each column:\n{df_raw.isnull().sum()}")

Number of null values in each column:
student_id                      0
gender                          0
study_time_hours                0
attendance_percent              0
sleep_hours                     0
parental_education            102
internet_access                 0
extracurricular_activities      0
part_time_job                   0
previous_grade                  0
final_exam_score                0
final_grade                     0
dtype: int64


We can see here that there are 102 null values listed for `parental_education`. We will need to inspect these further and transform them in the next section - **2. Transform**.

---